include/unifex/async_generator.hpp

```diff
@@ -0,0 +1,374 @@
+  using error_types = Variant<std::exception_ptr>;
+
+  static constexpr bool sends_done = false;
+
+  static constexpr auto blocking = blocking_kind::always_inline;
+
+  template <class R>
+  struct _op_state {
+    R r_;
+
+    friend void tag_invoke(tag_t<start>, _op_state& op) noexcept {
+      try {
+        set_value(std::move(op.r_), get_scheduler(op.r_));
+      } catch (...) {
+        set_error(std::move(op.r_), std::current_exception());
+      }
+    }
+  };
+
+  template <typename R>
+  friend _op_state<std::decay_t<R>>
+  tag_invoke(tag_t<connect>, _read_scheduler_sender, R&& r) {
+    return {std::forward<R>(r)};
+  }
+};
+
+struct _read_scheduler_t {
+  constexpr _read_scheduler_sender operator()() const noexcept { return {}; }
+};
+
+inline constexpr _read_scheduler_t read_scheduler{};
+
+template <typename Promise>
+struct reschedule_receiver {
+  std::optional<typename Promise::value_type> value_;
+  unifex::coro::coroutine_handle<Promise> genCoro_;
+
+  // set_value implies we are back on the correct scheduler => wake up the
+  // receiver with the value from the generator
+  void set_value() noexcept {
+    if (value_) {
+      unifex::set_value(std::move(*genCoro_.promise().receiverOpt_), *value_);
+    } else {
+      unifex::set_done(std::move(*genCoro_.promise().receiverOpt_));
+    }
+  }
+
+  // set_done and set_error is just a forward into the receiver
+  void set_done() noexcept {
+    unifex::set_done(std::move(*genCoro_.promise().receiverOpt_));
+  }
+
+  template <typename E>
+  void set_error(E&& err) noexcept {
+    unifex::set_error(
+        std::move(*genCoro_.promise().receiverOpt_), std::forward<E>(err));
+  }
+};
+
+// Note this is also used in final_suspend, since it handles moving
+// back to the consumer's scheduler, instead of using at_coroutine_exit
+// as in unifex::task<>
+template <typename T>
+class async_generator_yield_operation {
+public:
+  using value_type = std::remove_reference_t<T>;
+
+  async_generator_yield_operation(std::optional<value_type> value = {}) noexcept
+    : value_{std::move(value)} {}
+
+  bool await_ready() const noexcept { return false; }
+
+  template <typename Promise>
+  void await_suspend([[maybe_unused]] unifex::coro::coroutine_handle<Promise>
+                         genCoro) noexcept {
+    const auto& consumerSched = genCoro.promise().consumerSched_;
+    if (unifex::get_scheduler(genCoro.promise()) != consumerSched) {
+      genCoro.promise().rescheduleOpSt_ = unifex::connect(
+          unifex::schedule(consumerSched),
+          reschedule_receiver<Promise>{std::move(value_), genCoro});
+      unifex::start(*genCoro.promise().rescheduleOpSt_);
+      return;
+    }
+
+    if (value_) {
+      unifex::set_value(std::move(*genCoro.promise().receiverOpt_), *value_);
+    } else {
+      unifex::set_done(std::move(*genCoro.promise().receiverOpt_));
+    }
+  }
+
+  void await_resume() noexcept {}
+
+private:
+  std::optional<value_type> value_;
+};
```

Comment:
Below, I argue that you can rely on the generator generating values with code like
```
co_yield <expr>;
```
which means that `<expr>` survives across a suspend point. I think this means you can just store the address of that value here.
```suggestion
template <typename T>
class async_generator_yield_operation {
public:
  using value_type = std::remove_reference_t<T>;

  async_generator_yield_operation() noexcept
    : value_{nullptr} {}

  explicit async_generator_yield_operation(value_type& value) noexcept
    : value_{&value} {}

  bool await_ready() const noexcept { return false; }

  template <typename Promise>
  void await_suspend([[maybe_unused]] unifex::coro::coroutine_handle<Promise>
                         genCoro) noexcept {
    const auto& consumerSched = genCoro.promise().consumerSched_;
    if (unifex::get_scheduler(genCoro.promise()) != consumerSched) {
      genCoro.promise().rescheduleOpSt_ = unifex::connect(
          unifex::schedule(consumerSched),
          reschedule_receiver<Promise>{value_, genCoro});
      unifex::start(*genCoro.promise().rescheduleOpSt_);
      return;
    }

    if (value_) {
      unifex::set_value(std::move(*genCoro.promise().receiverOpt_), *value_);
    } else {
      unifex::set_done(std::move(*genCoro.promise().receiverOpt_));
    }
  }

  void await_resume() noexcept {}

private:
  value_type* value_;
};
```